enterprise/server/cmd/ci_runner/main.go

```diff
@@ -758,79 +784,152 @@ func setupGitRepo(ctx context.Context) error {
+	cfg := [][]string{
+		{"user.email", "ci-runner@buildbuddy.io"},
+		{"user.name", "BuildBuddy"},
+		{"advice.detachedHead", "false"},
 	}
-	tree, err := repo.Worktree()
-	if err != nil {
-		return status.WrapErrorf(err, "get worktree")
+	writeCommandSummary(ws.log, "Configuring repository...")
+	for _, kv := range cfg {
+		// Don't show the config output.
+		if err := git(ctx, io.Discard, "config", kv[0], kv[1]); err != nil {
+			return err
+		}
 	}
-	if err := tree.Clean(&git.CleanOptions{Dir: true}); err != nil {
-		return status.WrapErrorf(err, "clean")
+	return nil
+}
+
+func (ws *workspace) fetch(ctx context.Context, remoteURL string, branches []string) error {
+	if len(branches) == 0 {
+		return nil
 	}
-	remote, err := repo.Remote(git.DefaultRemoteName)
+	authURL, err := gitutil.AuthRepoURL(remoteURL, os.Getenv(repoUserEnvVarName), os.Getenv(repoTokenEnvVarName))
 	if err != nil {
-		return status.WrapErrorf(err, "get existing remote %q", git.DefaultRemoteName)
-	}
-	if err := fetchOrigin(ctx, remote); err != nil {
 		return err
 	}
-	checkoutOpts := &git.CheckoutOptions{
-		Hash:  gitplumbing.NewHash(*commitSHA),
-		Force: true, // remove local changes
+	remoteName := gitRemoteName(remoteURL)
+	writeCommandSummary(ws.log, "Configuring remote %q...", remoteName)
+	// Don't show `git remote add` command or the error message since the URL may
+	// contain the repo access token.
+	if err := git(ctx, io.Discard, "remote", "add", remoteName, authURL); err != nil && !isRemoteAlreadyExists(err) {
+		return status.UnknownErrorf("Command `git remote add %q <url>` failed.", remoteName)
 	}
-	if err := tree.Checkout(checkoutOpts); err != nil {
-		return status.WrapErrorf(err, "checkout %s", *commitSHA)
+	fetchArgs := append([]string{"fetch", "--force", remoteName}, branches...)
+	if err := git(ctx, ws.log, fetchArgs...); err != nil {
+		return err
 	}
 	return nil
 }
 
+type gitError struct {
+	error
+	Output string
+}
+
+func isRemoteAlreadyExists(err error) bool {
+	gitErr, ok := err.(*gitError)
+	return ok && strings.Contains(gitErr.Output, "already exists")
+}
+func isBranchNotFound(err error) bool {
+	gitErr, ok := err.(*gitError)
+	return ok && strings.Contains(gitErr.Output, "not found")
+}
+func isAlreadyUpToDate(err error) bool {
+	gitErr, ok := err.(*gitError)
+	return ok && strings.Contains(gitErr.Output, "up to date")
+}
+
+func git(ctx context.Context, out io.Writer, args ...string) *gitError {
+	var buf bytes.Buffer
+	w := io.MultiWriter(out, &buf)
+	printCommandLine(out, "git", args...)
+	if err := runCommand(ctx, "git", args, map[string]string{} /*=env*/, w); err != nil {
+		return &gitError{err, string(buf.Bytes())}
+	}
+	return nil
+}
+
+func writeCommandSummary(out io.Writer, format string, args ...interface{}) {
+	io.WriteString(out, ansiGray)
+	io.WriteString(out, "[BUILDBUDDY] ")
+	io.WriteString(out, fmt.Sprintf(format, args...))
+	io.WriteString(out, ansiReset)
+	io.WriteString(out, "\n")
```

Comment:
you have aurora.Sprintf(aurora.Red("\nAction failed: %s") below -- why not use that here?

Reply:
The ANSI codes that Aurora uses for gray aren't handled properly by our UI.